### PR TITLE
gobgpd: init at 2.26.0

### PR DESCRIPTION
**Pull request by higebu**
###### Motivation for this change

GoBGP is an open source BGP implementation. https://github.com/osrg/gobgp
Related to: https://github.com/NixOS/nixpkgs/pull/111083

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
    ```console
    $ nix path-info -S /nix/store/by6m9mkjfphj7q6cq5vx58jwj904dp8a-gobgpd-2.26.0
    /nix/store/by6m9mkjfphj7q6cq5vx58jwj904dp8a-gobgpd-2.26.0          15990496
    ```
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---
This change is [Reviewable](https://reviewable.io/reviews/nixos/nixpkgs/111316)

### CODE REVIEW

File: nixos/modules/services/networking/gobgpd.nix [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+          ];
+        }
+      '';
+    };
+  };
+
+  config = mkIf (cfg.enable) {
+    environment.systemPackages = [ cfg.package ];
+    environment.etc."gobgpd.conf".source =
+      format.generate "gobgpd.conf" cfg.settings;
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/gobgpd.nix:59]:
```suggestion
    environment.etc."gobgpd.conf".source = format.generate "gobgpd.conf" cfg.settings;
```

Comment (higebu) [reply nixos/modules/services/networking/gobgpd.nix]: Fixed it. Thank you.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 111316` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package blacklisted:
</summary>
<ul>
<li>tests.nixos-functions.nixos-test</li>
</ul>
</details>
<details>
<summary>
1 package built:
</summary>
<ul>
<li>gobgpd</li>
</ul>
</details>

**Comment by higebu** [comment]
Result of `nixpkgs-review pr 111316` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>gobgpd</li>
  </ul>
</details>

**Comment by higebu** [comment]
@SuperSandro2000 Can I do anything to merge this pr?

**Review by SuperSandro2000 — APPROVED** [review]
package LGTM

not sure about the module or the test.

**Comment by higebu** [comment]
@SuperSandro2000 Thank you for your approval.

**Comment by higebu** [comment]
@kalbasit Cloud you review this PR?

File: nixos/modules/services/networking/gobgpd.nix [hunk 2]
```diff
@@ -0,0 +1,73 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.services.gobgpd;
+  format = pkgs.formats.toml { };
+in {
+  options.services.gobgpd = {
+    enable = mkEnableOption "Whether to enable GoBGP Routing Daemon.";
```
Comment (aanderse) [inline nixos/modules/services/networking/gobgpd.nix:10]: You'll find that some of the text here is redundant. Please take a look at the [definition](https://github.com/NixOS/nixpkgs/blob/e2080b39e60b5d5113d6c226e49d946ce32b1a25/lib/options.nix#L97) of `mkEnableOption`.

File: nixos/modules/services/networking/gobgpd.nix [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    package = mkOption {
```
Comment (aanderse) [inline nixos/modules/services/networking/gobgpd.nix:12]: It appears we only have 1 version of this package and there are no variants or compile options exposed for it. Does this option provide any value?

File: nixos/modules/services/networking/gobgpd.nix [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                peer-as = 65002;
+              };
+            }
+          ];
+        }
+      '';
+    };
+  };
+
+  config = mkIf (cfg.enable) {
```
Comment (aanderse) [inline nixos/modules/services/networking/gobgpd.nix:56]: Please remove parenthesis.

File: nixos/modules/services/networking/gobgpd.nix [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    environment.systemPackages = [ cfg.package ];
+    environment.etc."gobgpd.conf".source = format.generate "gobgpd.conf" cfg.settings;
```
Comment (aanderse) [inline nixos/modules/services/networking/gobgpd.nix:58]: Some people strongly discourage placing configuration files under `/etc` unless the software absolutely demands it. The alternative being to specify the configuration file location in `ExecStart` instead, of course.

**Review by aanderse — CHANGES_REQUESTED** [review]
The module looks good, just a few minor suggestions.
Assuming I knew anything about this software (which I do not) I would *probably* say the test looks fine too.

:+1:

File: nixos/modules/services/networking/gobgpd.nix (hunk 2 above)
Comment (higebu) [reply nixos/modules/services/networking/gobgpd.nix:10]: I understand. Thank you!

File: nixos/modules/services/networking/gobgpd.nix (hunk 3 above)
Comment (higebu) [reply nixos/modules/services/networking/gobgpd.nix:12]: No. This option is not necessary. Thank you.

File: nixos/modules/services/networking/gobgpd.nix (hunk 4 above)
Comment (higebu) [reply nixos/modules/services/networking/gobgpd.nix]: Fixed. Thank you.

File: nixos/modules/services/networking/gobgpd.nix (hunk 5 above)
Comment (higebu) [reply nixos/modules/services/networking/gobgpd.nix]: Changed to generated configuration file path instead of under `/etc`.

**Comment by higebu** [comment]
@aanderse I fixed this module based on your review comments. Thank you.

**Review by aanderse — APPROVED** [review]
The module portion of this PR looks good to me. Great work! :+1:

**Comment by higebu** [comment]
@aanderse Thank you!

File: nixos/modules/services/networking/gobgpd.nix [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    systemd.services.gobgpd = {
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      description = "GoBGP Routing Daemon";
+      serviceConfig = {
+        Type = "notify";
+        ExecStartPre = "${pkgs.gobgpd}/bin/gobgpd -f ${confFile} -d";
+        ExecStart = "${pkgs.gobgpd}/bin/gobgpd -f ${confFile} --sdnotify";
+        ExecReload = "${pkgs.gobgpd}/bin/gobgpd -r";
+        DynamicUser = true;
```
Comment (mweinelt) [inline nixos/modules/services/networking/gobgpd.nix:59]: Can gobpgd sync its routes into the kernel routing table? If so, we'd need CAP_NET_ADMIN.

File: nixos/tests/gobgpd.nix [hunk 7]
```diff
@@ -0,0 +1,47 @@
+import ./make-test-python.nix ({ pkgs, ... } : {
+  name = "gobgpd";
+
+  meta = with pkgs.lib.maintainers; {
+    maintainers = [ higebu ];
+  };
+
+  nodes = {
+    node = { ... }: {
```
Comment (mweinelt) [inline nixos/tests/gobgpd.nix:9]: The test should likely have multiple nodes to establish bgp sessions amongst each other.

File: nixos/modules/services/networking/gobgpd.nix (hunk 6 above)
Comment (aanderse) [reply nixos/modules/services/networking/gobgpd.nix:59]: https://github.com/osrg/gobgp/blob/43da41ae63da75c9b38f5c4d9920af804b1e8d62/tools/contrib/centos/gobgpd.service 🤷‍♂️

File: pkgs/servers/misc/gobgpd/default.nix [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  buildFlagsArray = ''
+    -ldflags=
+    -s -w -extldflags '-static'
+  '';
+
+  subPackages = [ "cmd/gobgpd" ];
+
+  meta = with lib; {
+    description = "BGP implemented in Go";
+    homepage = "https://osrg.github.io/gobgp/";
```
Comment (mweinelt) [inline pkgs/servers/misc/gobgpd/default.nix:29]:
```suggestion
    homepage = "https://osrg.github.io/gobgp/";
    changelog = "https://github.com/osrg/gobgp/releases/tag/v${version}";
```

**Review by mweinelt — COMMENTED** [review]
Some open questions.

File: nixos/modules/services/networking/gobgpd.nix (hunk 6 above)
Comment (mweinelt) [reply nixos/modules/services/networking/gobgpd.nix:59]: If gobgp doesn't manipulate the FIB itself, but shelves that job out to Zebra, that would be fine.

Comment (higebu) [reply nixos/modules/services/networking/gobgpd.nix:59]: gobgpd cannot sync its routes into the kernel routing table. If we want to do it, we need zebra. For more details, see https://github.com/osrg/gobgp/blob/master/docs/sources/zebra.md

Comment (mweinelt) [reply nixos/modules/services/networking/gobgpd.nix:59]: That's fine then.

**Comment by mweinelt** [comment]
Sorry to mention this so late, but you need to split the changes into multiple commits.

- Package: `gobpgd: init at 2.26.0`
- Module: `nixos/gobpgd: init`
- Test: `nixos/tests/gobgpd: init`

**Comment by higebu** [comment]
@mweinelt Thank you for your comment. I fixed commits.

**Comment by mweinelt** [comment]
Looks awesome! Let's wait for the tests and then go for the merge! :fireworks: 

@GrahamcOfBorg test gobpgd

**Comment by mweinelt** [comment]
@GrahamcOfBorg build nixosTests.gobgpd

**Review by mweinelt — APPROVED** [review]

**Comment by higebu** [comment]
@mweinelt Thank you!

**Comment by mweinelt** [comment]
No, thank *you*! :man_dancing:

**Pull request merged**